File: platforms/android/app/src/main/java/io/foxbiz/shellular/browser/EmbeddedProxyService.java
package io.foxbiz.shellular.browser;

import android.content.Context;
import android.webkit.WebView;

import org.json.JSONArray;

import io.foxbiz.shellular.Callback;
import io.foxbiz.shellular.Service;

public class EmbeddedProxyService extends Service {

    private final EmbeddedProxyServer server = EmbeddedProxyServer.INSTANCE;

    public EmbeddedProxyService(Context context, WebView webView) {
        super(context, webView);
        server.setMainWebView(webView);
    }

    @Override
    public boolean exec(String action, JSONArray args, Callback callback) {
        switch (action) {
            case "responseStart":
                return handleResponseStart(args, callback);
            case "responseData":
                return handleResponseData(args, callback);
            case "responseEnd":
                return handleResponseEnd(args, callback);
            case "responseError":
                return handleResponseError(args, callback);
            default:
                callback.error("Unknown action: " + action);
                return false;
        }
    }

    private boolean handleResponseStart(JSONArray args, Callback callback) {
        try {
            String requestId = args.getString(0);
            int status = args.getInt(1);
            String statusText = args.getString(2);
            String headersJson = args.getString(3);

            server.sendResponseStart(requestId, status, statusText, headersJson);
            callback.success();
            return true;
        } catch (Exception e) {
            callback.error(e.getMessage());
            return false;
        }
    }

    private boolean handleResponseData(JSONArray args, Callback callback) {
        try {
            String requestId = args.getString(0);
            String chunk = args.getString(1);
            int index = args.getInt(2);

            server.sendResponseData(requestId, chunk, index);
            callback.success();
            return true;
        } catch (Exception e) {
            callback.error(e.getMessage());
            return false;
        }
    }

    private boolean handleResponseEnd(JSONArray args, Callback callback) {
        try {
            String requestId = args.getString(0);
            server.sendResponseEnd(requestId);
            callback.success();
            return true;
        } catch (Exception e) {
            callback.error(e.getMessage());
            return false;
        }
    }

    private boolean handleResponseError(JSONArray args, Callback callback) {
        try {
            String requestId = args.getString(0);
            String errorMessage = args.optString(1, "Unknown error");
            server.sendResponseError(requestId, errorMessage);
            callback.success();
            return true;
        } catch (Exception e) {
            callback.error(e.getMessage());
            return false;
        }
    }

    public boolean ensureServer(int port) {
        return server.startServer(port);
    }

    public void cleanupServer(int port) {
        server.stopServer(port);
    }

    public void cleanupAll() {
        server.cancelAllPending();
        server.stopAll();
    }

    public boolean hasServer(int port) {
        return server.hasServer(port);
    }

    @Override
    public void onDestroy() {
        cleanupAll();
    }
}

File: platforms/macos/shellular/EmbeddedProxyService.swift
final class EmbeddedProxyService: BaseService {

    private let server = EmbeddedProxyServer.shared

    override func exec(action: String, args: [Any], callback: Callback) {
        switch action {
        case "responseStart":
            handleResponseStart(args: args, callback: callback)
        case "responseData":
            handleResponseData(args: args, callback: callback)
        case "responseEnd":
            handleResponseEnd(args: args, callback: callback)
        case "responseError":
            handleResponseError(args: args, callback: callback)
        default:
            callback.error("Unknown action: \(action)")
        }
    }

    // MARK: - Response Handlers

    private func handleResponseStart(args: [Any], callback: Callback) {
        guard let requestId = args[safe: 0] as? String,
              let status = args[safe: 1] as? Int,
              let statusText = args[safe: 2] as? String,
              let headersJson = args[safe: 3] as? String else {
            callback.error("Invalid responseStart args")
            return
        }

        server.sendResponseStart(requestId: requestId, status: status, statusText: statusText, headersJson: headersJson)
        callback.success()
    }

    private func handleResponseData(args: [Any], callback: Callback) {
        guard let requestId = args[safe: 0] as? String,
              let chunk = args[safe: 1] as? String,
              let index = args[safe: 2] as? Int else {
            callback.error("Invalid responseData args")
            return
        }

        server.sendResponseData(requestId: requestId, chunkBase64: chunk, index: index)
        callback.success()
    }

    private func handleResponseEnd(args: [Any], callback: Callback) {
        guard let requestId = args[safe: 0] as? String else {
            callback.error("Invalid responseEnd args")
            return
        }

        server.sendResponseEnd(requestId: requestId)
        callback.success()
    }

    private func handleResponseError(args: [Any], callback: Callback) {
        guard let requestId = args[safe: 0] as? String else {
            callback.error("Invalid responseError args")
            return
        }
        let errorMessage = args[safe: 1] as? String ?? "Unknown error"
        server.sendResponseError(requestId: requestId, errorMessage: errorMessage)
        callback.success()
    }

    // MARK: - Server Lifecycle

    func ensureServer(port: Int) -> Bool {
        guard let webView = bridge?.webView else { return false }
        return server.startServer(port: port, webView: webView)
    }

    func cleanupServer(port: Int) {
        server.stopServer(port: port)
    }

    func cleanupAll() {
        server.cancelAllPending()
        server.stopAll()
    }

    func hasServer(port: Int) -> Bool {
        return server.hasServer(port: port)
    }
}
